enhancements/monitoring/alert-overrides.md

```diff
@@ -0,0 +1,470 @@
+}
+
+// Rule describes an alerting rule.
+// See Prometheus documentation: [alerting rules](https://www.prometheus.io/docs/prometheus/latest/configuration/alerting_rules/)
+// +k8s:openapi-gen=true
+type Rule struct {
+	Alert       string             `json:"alert"`
+	Expr        intstr.IntOrString `json:"expr"`
+	For         string             `json:"for,omitempty"`
+	Labels      map[string]string  `json:"labels,omitempty"`
+	Annotations map[string]string  `json:"annotations,omitempty"`
+}
+
+// AlertingRuleStatus is the status of an AlertingRule resource.
+type AlertingRuleStatus struct {
+	// observedGeneration is the last generation change you've dealt with
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// prometheusRule is the generated PrometheusRule for this AlertingRule
+	// +optional
+	PrometheusRule string `json:"prometheusRule,omitempty"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+
+// AlertRelabelConfig defines a set of relabel configs for alerts.
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +k8s:openapi-gen=true
+type AlertRelabelConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the desired state of this AlertRelabelConfig object.
+	Spec AlertRelabelConfigSpec `json:"spec"`
+}
+
+// AlertRelabelConfigsSpec is the desired state of an AlertRelabelConfig resource.
+// +k8s:openapi-gen=true
+type AlertRelabelConfigSpec struct {
+	Configs []RelabelConfig `json:"configs"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AlertRelabelConfigList is a list of AlertRelabelConfigs.
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +k8s:openapi-gen=true
+type AlertRelabelConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// items is a list of AlertRelabelConfigs.
+	Items []*AlertRelabelConfig `json:"items"`
+}
+
+// LabelName is a valid Prometheus label name which may only contain ASCII
+// letters, numbers, and underscores.
+// +kubebuilder:validation:Pattern:="^[a-zA-Z_][a-zA-Z0-9_]*$"
+type LabelName string
+
+// RelabelConfig allows dynamic rewriting of label sets for alerts.
+// +k8s:openapi-gen=true
+type RelabelConfig struct {
+	// sourceLabels select values from existing labels. Their content is
+	// concatenated using the configured separator and matched against the
+	// configured regular expression for the replace, keep, and drop actions.
+	SourceLabels []LabelName `json:"sourceLabels,omitempty"`
```

Comment:
Is this optional? Does it have a minimum length?

Reply:
It is optional depending on the action, yes. I've marked it as such. Relabelling is an advanced feature, and basically every field here is optional or required based on the action chosen. I've tried to expand the docs for each with more information though.